Gems/EMotionFX/Code/Source/Editor/Plugins/Ragdoll/JointLimitRotationManipulators.cpp

```diff
@@ -0,0 +1,162 @@
+            m_rotationManipulators.SetLocalOrientation(GetLocalOrientation());
+        }
+    }
+
+    void JointLimitRotationManipulators::Teardown()
+    {
+        PhysicsSetupManipulatorRequestBus::Handler::BusDisconnect();
+        AZ::TickBus::Handler::BusDisconnect();
+        m_rotationManipulators.Unregister();
+    }
+
+    void JointLimitRotationManipulators::ResetValues()
+    {
+        if (m_physicsSetupManipulatorData.HasJointLimit())
+        {
+            BeginEditing();
+            GetLocalOrientation() = AZ::Quaternion::CreateIdentity();
+            FinishEditing();
+            Refresh();
+        }
+    }
+
+    void JointLimitRotationManipulators::OnManipulatorMoved(const AZ::Quaternion& rotation)
+    {
+        m_rotationManipulators.SetLocalOrientation(rotation);
+        if (m_physicsSetupManipulatorData.HasJointLimit())
+        {
+            GetLocalOrientation() = rotation;
+        }
+        m_physicsSetupManipulatorData.m_jointLimitWidget->InvalidateValues();
+    }
+
+    void JointLimitRotationManipulators::BeginEditing()
+    {
+        if (!m_commandGroup.IsEmpty())
+        {
+            return;
+        }
+        m_commandGroup.SetGroupName("Adjust joint limit");
+        const AZ::u32 actorId = m_physicsSetupManipulatorData.m_actor->GetID();
+        const AZStd::string& nodeName = m_physicsSetupManipulatorData.m_node->GetNameString();
+        CommandAdjustJointLimit* command = aznew CommandAdjustJointLimit(actorId, nodeName);
+        m_commandGroup.AddCommand(command);
+        command->SetOldJointConfiguration(m_physicsSetupManipulatorData.m_jointConfiguration);
+    }
+
+    void JointLimitRotationManipulators::FinishEditing()
+    {
+        if (m_commandGroup.IsEmpty())
+        {
+            return;
+        }
+
+        if (CommandAdjustJointLimit* command = azdynamic_cast<CommandAdjustJointLimit*>(m_commandGroup.GetCommand(0)))
+        {
+            command->SetJointConfiguration(m_physicsSetupManipulatorData.m_jointConfiguration);
+        }
+        AZStd::string result;
+        CommandSystem::GetCommandManager()->ExecuteCommandGroup(m_commandGroup, result);
+        m_commandGroup.Clear();
+    }
+
+    void JointLimitRotationManipulators::OnTick([[maybe_unused]] float delta, [[maybe_unused]] AZ::ScriptTimePoint timePoint)
+    {
+        const AzFramework::CameraState cameraState = AzToolsFramework::GetCameraState(m_viewportId);
+        m_rotationManipulators.RefreshView(cameraState.m_position);
+    }
+
+    void JointLimitRotationManipulators::OnUnderlyingPropertiesChanged()
+    {
+        Refresh();
+    }
+
+    AZ::Quaternion& JointLimitRotationManipulators::GetLocalOrientation()
+    {
+        return m_jointLimitFrame == JointLimitFrame::Parent
+            ? m_physicsSetupManipulatorData.m_jointConfiguration->m_parentLocalRotation
+            : m_physicsSetupManipulatorData.m_jointConfiguration->m_childLocalRotation;
+    }
```

Comment:
minor: this can call the const version of `GetLocalOrientation` to only have one implementation.

Reply:
I actually tried that first but I don't think it works because the functions both take no arguments, so if you try to call the other one it has no way to know which one to call and winds up calling itself recursively. And it's such a short implementation that I didn't think it was too bad to do it this way.

Reply:
mmmmm, casting `*this` to const should work because it has the context of what function to call. It's a bit verbose, but it should work. If you're curious give it a go, but if it's not a good moment to do the change don't worry about it.

````
AZ::Quaternion& JointLimitRotationManipulators::GetLocalOrientation()
{
    return const_cast<AZ::Quaternion&>(static_cast<const JointLimitRotationManipulators&>(*this).GetLocalOrientation());
}
````

Reply:
Classic - https://stackoverflow.com/a/123995

Reply:
Updated